keps/sig-apps/0008-sidecarcontainer.md

```diff
@@ -0,0 +1,108 @@
+reviewers:
+  - TBD
+approvers:
+  - TBD
+editor: TBD
+creation-date: 2018-05-14
+last-updated: 2018-05-14
+status: provisional
+---
+
+# Sidecar Containers
+
+## Table of Contents
+
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [Implementation Details/Notes/Constraints [optional]](#implementation-detailsnotesconstraints-optional)
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+* [Alternatives](#alternatives-optional)
+
+## Summary
+
+To solve the issue of sidecars not being well supported we can create a new class of container in the Pod Spec called sidecarContainers which are handled like normal containers but are not considered when evaluating overall pod status. This will solve the problem of Jobs not completing/failing when a sidecar container never exits due to it having no awareness of the 'primary' container.
+
+## Motivation
+
+SideCar containers have always been used in some ways but just not formally identified as such, with the introduction of Jobs and Cronjobs it became clear that they're not handled very well. If you have a Job with two containers one of which is actually doing the main processing of the job and the other is just facilitating it, you encounter a problem when the main process finishes; your sidecar container will carry on running so the job will never finish. This only way around this is to build some kind of logic into the sidecar so that it will exit when the other container exits. This can be quite difficult if you're using a sidecar provided by a third party and results in some nasty hacks being implemented to achieve this.
+
+### Goals
+
+Solve issue [25908](https://github.com/kubernetes/kubernetes/issues/25908) so that sidecars can be easily handled without needing modifications to the container itself
+
+### Non-Goals
+
+Changing the way jobs are handled
+
+## Proposal
+
+Create a new class of containers called sidecarContainers, this will be an additional field to the Pod Spec.
+e.g
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: myapp-pod
+  labels:
+    app: myapp
+spec:
+  containers:
+  - name: myapp
+    image: myapp
+    command: ['do something']
+  sidecarContainers:
+  - name: sidecar
+    image: sidecar-image
+    command: ["do something to help my app"]
+
+```
+These sidecar containers will be handled the same way as a normal container is almost all respects apart from the fact that they will not be evaluated when determining the pod status.
+In a job for example this means that you can have as many sidecars as you like and they will all get terminated when the normal containers all Complete.
+This model assumes that the sidecars are not intended to serve traffic and your app will only pass readiness check if the sidecar is successfully running.
```

Comment:
Can you be more specific about "serve traffic"? I think what you mean is "primary workload"? 

"serve traffic"  to me indicates that your sidecar can't be anything that serves network traffic like https proxy sidecar, envoy, etc.

Reply:
yeah that is a bit vague, I'm not sure how to phrase it, something along of the lines of it shouldn't be the "main process" or is not reliant on other containers to run, it's just something that facilitates your actual app running.